src/openvic-simulation/dataloader/NodeTools.hpp

```diff
@@ -18,6 +22,28 @@ namespace OpenVic {
 
 	namespace NodeTools {
 
+		template<typename Fn, typename Return = void, typename ...Args>
+		concept Functor = requires(Fn&& fn, Args&& ...args) {
+			{ std::invoke(std::forward<Fn>(fn), std::forward<Args>(args)...) } -> std::same_as<Return>;
+		};
+
+		template<typename Fn, typename Return = void, typename ...Args>
+		concept FunctorConvertible = requires(Fn&& fn, Args&& ...args) {
+			{ std::invoke(std::forward<Fn>(fn), std::forward<Args>(args)...) } -> std::convertible_to<Return>;
+		};
+
+		template<typename Fn, typename ...Args>
+		concept Callback = Functor<Fn, bool, Args...>;
+
+		template<typename Fn, typename ...Args>
+		concept NodeCallback = Callback<Fn, ast::NodeCPtr>;
+
+		template<typename Fn, typename ...Args>
+		concept KeyValueCallback = Callback<Fn, std::string_view, ast::NodeCPtr>;
+
+		template<typename Fn, typename ...Args>
+		concept LengthCallback = Functor<Fn, std::size_t, std::size_t>;
```

Comment:
Do these still need the `Args` template parameter?

Reply:
Yeah it really doesn't